torch/optim/lazy_adam.py

```diff
@@ -0,0 +1,93 @@
+        eps (float, optional): term added to the denominator to improve
+            numerical stability (default: 1e-8)
+        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+
+    .. _Adam\: A Method for Stochastic Optimization:
+        https://arxiv.org/abs/1412.6980
+    """
+
+    def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8):
+        defaults = dict(lr=lr, betas=betas, eps=eps)
+        super(LazyAdam, self).__init__(params, defaults)
+
+    def step(self, closure=None):
+        """Performs a single optimization step.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            for p in group['params']:
+                if p.grad is None:
+                    continue
+                grad = p.grad.data
+                if not grad.is_sparse:
+                    raise RuntimeError('LazyAdam does not support dense gradients, please consider using Adam instead')
+
+                state = self.state[p]
+
+                # State initialization
+                if len(state) == 0:
+                    state['step'] = 0
+                    # Exponential moving average of gradient values
+                    state['exp_avg'] = p.data.new().resize_as_(p.data).zero_()
+                    # Exponential moving average of squared gradient values
+                    state['exp_avg_sq'] = p.data.new().resize_as_(p.data).zero_()
+
+                state['step'] += 1
+
+                grad = grad.coalesce()  # the update is non-linear so indices must be unique
+                grad_indices = grad._indices()
+                grad_values = grad._values()
+                grad_values_sq = grad_values.pow(2)
+                size = torch.Size([x for x in grad.size()])
+
+                def make_sparse(values):
+                    constructor = type(p.grad.data)
+                    if grad_indices.dim() == 0 or values.dim() == 0:
+                        return constructor()
+                    return constructor(grad_indices, values, size)
+
+                exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
+                beta1, beta2 = group['betas']
+
+                # Decay the first and second moment running average coefficient
+                #      old <- b * old + (1 - b) * new
+                # <==> old += (1 - b) * (new - old)
+                old_exp_avg_values = exp_avg._sparse_mask(grad)._values()
+                exp_avg.add_(make_sparse(grad_values.sub_(old_exp_avg_values).mul_(1 - beta1)))
+                old_exp_avg_sq_values = exp_avg_sq._sparse_mask(grad)._values()
+                exp_avg_sq.add_(make_sparse(grad_values_sq.sub_(old_exp_avg_sq_values).mul_(1 - beta2)))
```

Comment:
same as above

Reply:
This should be okay, as the `grad_values_sq` is a new tensor.